Pr fix ots sample

pirun opened the pull request:
This PR brings several fix for Bluetooth OTS service and sample.

- Check length and offset according to object size and properties.
- Fix peripheral sample object data index error. object array index should be calculated by offset of selected ID and MIN object ID.
- Central should check object proprieties before read or write.

alwa-nordic commented on include/zephyr/bluetooth/services/ots.h:
```diff
@@ -902,6 +902,17 @@ struct bt_ots_client_cb {
+ *
+ *  @return int            0 if success, ERRNO on failure.
+ */
+int bt_ots_client_unregister(uint8_t index);
```
There is no way for the user of this API to get the index. I would prefer `bt_ots_client_unregister(struct bt_ots_client *ots_inst)`. Alternatively, change `bt_ots_client_register` so that non-negative return values are the assigned index. Or add a `bt_ots_get_client_handle`/`index`.

pirun replied:
Currently the OTS client instance number is hardcoded as 1. So user should just assign 0 to the index.
https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/services/ots/ots_client.c#L35
I have no idea if the original maintainer plan to do multi instances implementation.

alwa-nordic replied:
> Currently the OTS client instance number is hardcoded as 1. So user should just assign 0 to the index. https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/services/ots/ots_client.c#L35 I have no idea if the original maintainer plan to do multi instances implementation.

The API and the current implementation are different things. We have to future-proof the API.

The `index`-approach does not make sense in this API. We can document that `index` must always be `0`, and other values are RFU. But that feels unwise to me because we are heading down a path where the user must keep track of the indexes. And the indexes should not be needed, since the `ots_inst`s are unique (right?). It smells to me of a leaked implementation detail - that the implementation is storing `ots_inst` pointers in an array.

I still suggest changing the prototype to `bt_ots_client_unregister(struct bt_ots_client *ots_inst)`. Another possibility is to have a `bt_ots_client_unregister_all(void)` instead.

pirun committed:
Bluetooth: OTS - Object Write check parameter properly

Check sum of len and offset with properties.
Check offset with obj cur size.

Signed-off-by: Pirun Lee <pirun.lee@nordicsemi.no>

pirun committed:
Bluetooth: ots: fix sample data operation error

Peripheral: Fix wrong objects index
Central: Fix client not working after re-connect.
         Check object properties before read/write.

Signed-off-by: Pirun Lee <pirun.lee@nordicsemi.no>

jhedberg commented on samples/bluetooth/central_otc/src/main.c:
```diff
@@ -168,14 +175,14 @@ static void configure_button_irq(const struct gpio_dt_spec btn)
 	}
 
 	ret = gpio_pin_configure_dt(&btn, GPIO_INPUT);
-	if (ret != 0) {
+	if (ret) {
```
Why are you making these changes? Even though the Bluetooth stack doesn't fall under the safety certification scope, we should still try to follow the Zephyr coding guidelines which strive for MISRA compliance. One of the requirements is that integers should not be used as booleans in conditional statements.

pirun replied:
I saw many of Zephyr sample use integers that way. Trying to align to those sample.

alwa-nordic replied:
@pirun I suggest we remove those changes to expedite the merging of this PR. It makes the diff smaller as well, which is nice.

pirun replied:
@jhedberg @alwa-nordic Agreed. Thanks for your opinions.

jhedberg replied:
Yes, I'm aware of inconsistencies. There's a lot of code in the tree that predates the safety certification related coding guidelines. A lot of that code hasn't been updated, and then there's probably new code that took example from the old code. FWIW, even if you'd have had these changes in this PR you shouldn't mix them into a commit that contains an unrelated functional change, rather style changes, etc should be in separate commits.

jhedberg replied:
For reference, I'm talking about MISRA C:2012 Rule 14.4. Here are some PRs which fix code related to it: #48070 #48082 #48065 #40415

pirun replied:
I'm wondering if it is okay to add additional commit to fix style in this PR?

pirun committed:
sample: ots: comply to coding guidelines MISRA C:2012 Rule 14.4

The controlling expression of an if statement
and the controlling expression of an iteration-statement shall have
essentially Boolean type.

Signed-off-by: Pirun Lee <pirun.lee@nordicsemi.no>
